### ISSUE TITLE
Reindex times out for a big graph

### ISSUE THREAD

**Issue by te-en** [JanusGraph/janusgraph#1389]
Observed on JanusGraph 0.3.1 with Cassandra as the storage backend and Solr as the index backend.

My graph has over 14 million vertices. I added a new mixed index and I wanted to run the reindex operation as described in the official documentation ([Executing a Reindex job on JanusGraphManagement](https://docs.janusgraph.org/latest/index-admin.html#_executing_a_reindex_job_on_janusgraphmanagement)). I made sure all indexed keys are in a status which permits reindexing by calling:
```
ManagementSystem.awaitGraphIndexStatus(graph, indexName).status(SchemaStatus.REGISTERED, SchemaStatus.ENABLED)
```
But whenever I try to execute the reindex job (from a Java program or from the Gremlin console):
```
mgmt.updateIndex(mgmt.getGraphIndex(indexName), SchemaAction.REINDEX).get();
```
it fails after approximately 3 minutes with the following error:
```
2019-01-19 00:34:31.723 ERROR    NONE 7 --- [       Thread-9] o.j.d.k.s.StandardScannerExecutor        : Exception occurred during job execution: {}
org.janusgraph.diskstorage.TemporaryBackendException: Timed out waiting for next row data - storage error likely
        at org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor.run(StandardScannerExecutor.java:155) [janusgraph-core-0.3.1.jar!/:?]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]
```
followed by this repeated a few times:
```
2019-01-19 00:34:31.827 ERROR    NONE 7 --- [      Thread-10] o.j.d.k.s.StandardScannerExecutor        : Data-pulling thread interrupted while waiting on queue or data
java.lang.InterruptedException: null
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.reportInterruptAfterWait(AbstractQueuedSynchronizer.java:2014) ~[?:1.8.0_191]
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2048) ~[?:1.8.0_191]
        at java.util.concurrent.LinkedBlockingQueue.put(LinkedBlockingQueue.java:350) ~[?:1.8.0_191]
        at org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor$DataPuller.run(StandardScannerExecutor.java:340) [janusgraph-core-0.3.1.jar!/:?]
```

I ran identical code on a smaller graph (around 30 thousand vertices) and it succeeded.

I also found this thread: https://groups.google.com/forum/#!msg/janusgraph-users/9KZG8y-AdNI/lI_HubxiCQAJ - which suggests I am not the only one facing this sort of problems for big graphs.

**Comment by krlohnes** [comment]
I also ran into this, I'm wondering if you figured out anything about this @te-en
```
gremlin> mgmt.updateIndex(mgmt.getGraphIndex(indexName), SchemaAction.REINDEX).get()
04:48:36 WARN  org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor  - Data pulling thread [1] did not terminate. Forcing termination
04:48:36 ERROR org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor  - Data-pulling thread interrupted while waiting on queue or data
java.lang.InterruptedException
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.reportInterruptAfterWait(AbstractQueuedSynchronizer.java:2014)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2048)
        at java.util.concurrent.LinkedBlockingQueue.put(LinkedBlockingQueue.java:350)
        at org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor$DataPuller.run(StandardScannerExecutor.java:335)
04:48:36 WARN  org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor  - Data pulling thread [2] did not terminate. Forcing termination
04:48:36 ERROR org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor  - Data-pulling thread interrupted while waiting on queue or data
java.lang.InterruptedException
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.reportInterruptAfterWait(AbstractQueuedSynchronizer.java:2014)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2048)
        at java.util.concurrent.LinkedBlockingQueue.put(LinkedBlockingQueue.java:350)
        at org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor$DataPuller.run(StandardScannerExecutor.java:335)
==>org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScanMetrics@3d104c9b
```

**Comment by te-en** [comment]
@krlohnes I didn't find any good solution, just an ugly workaround. I explicitly reset all indexed properties for all vertices, so something like:
```
.has('description').property('description', values('description'))
```
Afterwards all descriptions could be found in Solr.

By the way I needed to utilise another mixed index to traverse my entire graph, otherwise I was running into `OutOfMemoryError`.

**Comment by narayana1043** [comment]
i also had this issue but i had it on the egde

https://github.com/JanusGraph/janusgraph/issues/1803

**Comment by twilmes** [comment]
I recently ran into this issue using the non-MapReduce reindexing method with the CQL storage adapter. I traced the OOM back to paging behavior of the Datastax Cassandra driver, though I wouldn't consider it a driver bug. JG is executing this scan query to scan the full key range in the storage layer and then iterating through it.
```
return Try.of(() -> new CQLResultSetKeyIterator(
                query,
                this.getter,
                this.session.execute(this.getKeysAll.bind()
                        .setBytes(SLICE_START_BINDING, query.getSliceStart().asByteBuffer())
                        .setBytes(SLICE_END_BINDING, query.getSliceEnd().asByteBuffer())
                        .setFetchSize(this.storeManager.getPageSize())
                        .setConsistencyLevel(getTransaction(txh).getReadConsistencyLevel()))))
                .getOrElseThrow(EXCEPTION_MAPPER);
```
The OOM issue arises because the Cassandra driver keeps information about each page it retrieves from Cassandra in memory and the driver API allows one to retrieve this information. This is fine if you're not attempting a full scan of the key range. Have you guys run into this same issue with the standalone MapReduce indexer? I don't believe it will exhibit this same behavior due to how the CassandraRecordReader owrks. Having said that, I think a fix for this is to add an intermediate level of paging on the JG side so that we do not keep a single iterator open that eventually OOMs due to its internal book keeping.

**Comment by mad** [comment]
@twilmes may be submit a issue here https://datastax-oss.atlassian.net/projects/JAVA/issues/ ?

**Comment by twilmes** [comment]
@mad thanks for the link. When I started looking into it, I thought perhaps it was a driver bug but upon closer inspection retrieval of the full page history is supported at the Cassandra driver API level so it's working as expected. In reality, I think JG's use of this paging support to scan an entire token range which on large graphs could likely be 10's to 100's of GBs (though at that point, a MapReduce job is a better idea) is not a great fit for this driver access pattern and not how it was intended to be used so I determined it's a quirk of our access patterns versus a driver bug.

**Comment by mad** [comment]
@twilmes I think is root cause is https://github.com/JanusGraph/janusgraph/issues/1750 because reindex just timeout but not OOM. What do you think?

**Comment by twilmes** [comment]
Hey @mad. I think you're right. It appears there are a few issues with reindexing. I'll open a new issue that describes this specific problem and update my PR.
